### PR TITLE
Mask non-finite floats properly

### PR DESCRIPTION
**Pull request by jl-wynen**
The old implementation does not work when the masked data contains non-finite floats (`NaN` or `inf`).

### CODE REVIEW

File: dataset/variable_reduction.cpp [hunk 1]
```diff
@@ -15,14 +16,10 @@
...
 
-namespace {
-Variable zero_like(const Variable &var) { return {var, Dimensions{}}; }
-} // namespace
-
 Variable sum(const Variable &var, const Dim dim, const Masks &masks) {
   if (const auto mask_union = irreducible_mask(masks, dim);
       mask_union.is_valid()) {
-    return sum(where(mask_union, zero_like(var), var), dim);
+    return sum(where(mask_union, zeros_like(var), var), dim);
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply dataset/variable_reduction.cpp:22]: The function is actually `zero_like`, i.e. it makes a scalar. I renamed it. But that does not work well with the actual `zeros_like` which only exists in Python. I'm changing the name back.

**Comment by jl-wynen** [comment]
The change causes a problem when histogramming binned data.
The call to `histogram(a, bin_edges)` in https://github.com/scipp/scipp/blob/b8e1174651a674ce33890efee9755ef2517ce831/dataset/test/bins_test.cpp#L168 throws
```
Binary operations such as '+' with binned data are only supported with dtype=VariableView, got dtype=DataArrayView. See https://scipp.github.io/user-guide/binned-data/computation.html#Event-centric-arithmetic for equivalent operations for binned (event) data.
```
I am unsure how to fix this. Is this a problem with `where` that can be fixed or do we have to avoid using `where`?

**Comment by SimonHeybrock** [comment]
Hmm, can you use the result of `where` to replace the `data` of the input data array (after making a shallow copy)?

**Comment by SimonHeybrock** [comment]
An alternative would be to define `where` for `DataArray`, similar to how we have `+` defined for the 4 combinations of `DataArray` and `Variable` inputs (I suppose the condition should only accept variables).

**Comment by SimonHeybrock** [comment]
The error is not what I initially thought, I think it is `where` that is failing directly. Looking into it...

**Comment by jl-wynen** [comment]
Is this all or did I miss something?

File: variable/creation.cpp [hunk 2]
```diff
@@ -62,4 +62,6 @@ Variable special_like(const Variable &prototype, const FillValue &fill) {
   throw std::runtime_error("Unsupported fill value.");
 }
 
+Variable zero_like(const Variable &var) { return {var, Dimensions{}}; }
```
Comment (SimonHeybrock) [inline variable/creation.cpp:65]: Maybe a docstring to explain that this is singular/scalar intentionally and should *not* be refactored to to what `zeros_like` does?

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**